### DISCUSSION TITLE
Basic question about the Delay node

### DISCUSSION THREAD

**Discussion by lachioma** [bonsai-rx/bonsai#1635] [Q&A] [answered]
Hi all

I have a Bonsai workflow with many separate sequences of operators and branches. One of these sequences (at the bottom of the workflow) contains a `Delay` node. When the program ends up in this `Delay` node, does it just wait here, without running the other sequences?
How could I set some kind of delay or timer that delays one sequence while allowing the program to run the other sequences?

Thank you for any feedback!

**Comment by bruno-f-cruz** [comment, marked as answer]
If the sequences are not dependent on each other they should run asynchronously and in parallel. Feel free to attach the workflow if you want more feedback on the specific pattern.

**Comment by lachioma** [comment]
Thank you Bruno!

In my workflow, I acquire images through a camera that is triggered via an Arduino.
At the bottom of my workflow there is the following sequence:

![image](https://github.com/bonsai-rx/bonsai/assets/29898879/a1fab363-30d7-416f-93f5-d245f513fbf6)

Basically, by pressing a key, I tell Arduino to stop triggering frames from the camera (via `SerialStringWrite`) and give a delay (0.5 sec using `Delay`) to allow the part of workflow located above this sequence to get and save the last image (or images) that might still be triggered after the key press. Finally the whole workflow stops.
I asked the original question because the video file often has 1 or few frames less than the number of Arduino trigger pulses, but I am confident that there are no dropped frames during the acquisition. So these "missed" frames have to be either at the beginning or at the end of the video recording. So my hypothesis was that if Bonsai would just wait in `Delay` without executing the other sequences, this might result in the final frames not being saved.

**Reply by bruno-f-cruz** [reply]
Attach your workflow. I suspect your problem might be at the start of the sequence not at the end.

**Reply by lachioma** [reply]
Here is my workflow: https://github.com/lachioma/bonsai_record_cameraFLIR_viaArduino/blob/main/record_cameraFLIR_viaArduino.bonsai

Just some comments to help you read through the workflow:
- the camera is controlled via an Arduino. Bonsai sends a '1' to Arduino every 0.2 seconds, Arduino triggers the camera frames as long as keeps on receiving these ones.
- The frame ID is overlaid on the camera images before saving into avi.
- A csv file with frame ID and timestamps is also saved.
- The acquisition stops by pressing a button (F8).

Thank you very much Bruno for looking into it!!

**Reply by bruno-f-cruz** [reply]
You must allow the camera to start before sending triggers to the camera. Try to implement the following logic in your workflow:
1. Create the camera object (SpinnakerCapture).
2. Create an Arduino object or serial object (CreateSerial)
3. Wait an arbitrary amount of time. Start with a couple of seconds just in case.
4. Start sending trigger messages. 

This strategy allows all objects to be created BEFORE you request a frame to be acquired.

**Reply by lachioma** [reply]
Okay, thank you very much!
I implemented your feedback in the this way, placing the following sequences after the camera object:

![image](https://github.com/bonsai-rx/bonsai/assets/29898879/34baeefe-ad3c-492d-8f70-b18406bda34a)

The delay works. I'll run some tests and post something here in case I still get dropped/missed frames.

**Reply by bruno-f-cruz** [reply]
I dont see the code on the camera initialization. Can you include the workflow? Keep in mind that if you want to increase the chances of someone helping you in the discussions, you should always include your workflow (read #864). Otherwise, this might turn into a long back-and-forth. Thanks!

**Reply by lachioma** [reply]
Sorry, I did not include the code on the camera because it is the same as posted in the [link ](https://github.com/lachioma/bonsai_record_cameraFLIR_viaArduino/blob/main/record_cameraFLIR_viaArduino.bonsai) above. I just placed this code before the Arduino object. (Next times I will always include screenshots instead of a link to the .bonsai file).

Here is the screenshot with the part of the workflow about the camera, followed by the Arduino sequences.

![image](https://github.com/bonsai-rx/bonsai/assets/29898879/8dcc2afe-8e9d-4fe6-8bae-992750d78082)
![image](https://github.com/bonsai-rx/bonsai/assets/29898879/2fc473fa-c91a-4513-a653-280b0ef21298)
